text/0000-gen-fn.md

```diff
@@ -0,0 +1,789 @@
+    (for ([x xs])
+      (when (odd? x)
+        (yield (* 2 x))))))
+
+(define g (odd-dup '(1 2 3 4 5)))
+(= (g) 2)
+(= (g) 6)
+(= (g) 10)
+```
+
+Note that because of the expressive power of [`call/cc`][racket-callcc] (and
+continuations in general), generators can be written in Racket as a normal
+library.
+
+[racket-callcc]: https://docs.racket-lang.org/reference/cont.html
+[racket-generators]: https://docs.racket-lang.org/reference/Generators.html
+
+## Haskell, Idris, Clean, etc.
+
+In [Haskell][] (and in similar languages such as [Idris][idris-lang],
+[Clean][clean-lang], etc.), all functions are lazy unless specially annotated.
+Consequently, Haskell does not need a special `yield` operator. Any function
+can be a generator by recursively building a list of elements that will be
+lazily returned one at a time. E.g.:
+
+```haskell
+oddDup :: (Integral x) => [x] -> [x]
+oddDup [] = []
+oddDup (x:xs)
+  | odd x = x * 2 : oddDup xs
+  | otherwise = oddDup xs
+
+main :: IO ()
+main = putStrLn $ show $ take 5 $ oddDup [1..20]
+```
+
+[haskell]: https://www.haskell.org/
+[clean-lang]: https://wiki.clean.cs.ru.nl/Clean
+[idris-lang]: https://www.idris-lang.org/
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Keyword
+
+Should we use `iter` as the keyword, as we're producing `Iterator`s?
+We could use `gen` as proposed in this RFC and later extend its abilities to more powerful generators.
+
+[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=efeacb803158c2ebd57d43b4e606c0b5)
+
+```rust
+#![feature(generators)]
+#![feature(iter_from_generator)]
+
+fn main() {
+    let mut it = std::iter::from_generator(|| {
+        yield 1
+    });
+
+    assert_eq!(it.next(), Some(1));
+    assert_eq!(it.next(), None);
+    it.next(); // panics
+}
+```
+
+## Panicking
+
+What happens when `Iterator::next` is called again on a `#[rustc_gen]` block that panicked? Do we need to poison the iterator?
+
+## Contextual keyword
+
+Popular crates (like `rand`) have methods called [`gen`][Rng::gen]. If we forbid those, we are forcing those crates to make a major version bump when they update their edition, and we are requiring any users of those crates to use `r#gen` instead of `gen` when calling that method.
+
+We could choose to use a contextual keyword and only forbid `gen` in:
+
+* bindings
+* field names (due to destructuring bindings)
+* enum variants
+* type names
+
+This should avoid any parsing issues around `gen` followed by `{` in expressions.
+
+[Rng::gen]: https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen
+
+## `Iterator::size_hint`
+
+Should we try to compute a conservative `size_hint`? This will reveal information from the body of a generator,
+but at least for simple cases users will likely expect `size_hint` to not just be the default.
+
+## Implement other `Iterator` traits.
+
+Is there a possibility for implementing traits like `DoubleEndedIterator`, `ExactSizeIterator` at all?
```

Comment:
Musing: If we had fully-general RTN we could possibly allow implementing traits on them:

```rust
gen fn foo() -> i32 { ... }
impl ExactSizeIterator for foo() {}
```

We could also allow a single `impl Iterator for foo()` block to override `size_hint`.

It doesn't solve the problem of implementing `DoubleEndedIterator` with a generator though. That sounds more like we need a DoubleEndedGenerator with a `resume_back` method, with an attribute and compiler analysis to implement it.